protocol/lavasession/single_provider_session.go

```diff
@@ -184,6 +195,22 @@ func (sps *SingleProviderSession) PrepareSessionForUsage(ctx context.Context, cu
 	return nil
 }
 
+func (sps *SingleProviderSession) validateAndAddBadgeUsedCU(currentCU uint64, maxCu uint64, badgeUserEpochData *ProviderSessionsEpochData) error {
```

Comment:
Missing implementation of onSessionFailure, when a session fails you need to substruct the CU as the badge didn't use the cu.

Reply:
onSessionFailure logic is implemented here: 3933fc142a1f6ae0b75c183c615ded94285e501d
Also, I have written a unit test for onSessionFailure on Badge sessions to validate CU is properly subtracted from badge user epoch data: 0f0c83576cbf21e1ca98cf116f73c9761fe62193